File: Sources/ChorusCore/Audio/JitterBuffer.swift
import Foundation

/// Holds received PCM briefly so Wi‑Fi jitter does not turn into audible gaps.
@MainActor
public final class AudioJitterBuffer {
    public struct Chunk {
        public let header: AudioChunkHeader
        public let pcm: Data
    }

    private let targetDuration: TimeInterval
    private let sampleRate: Double
    private var pending: [UInt64: Chunk] = [:]
    private var nextSequence: UInt64 = 0
    private var bufferedSamples = 0
    private var started = false

    public init(sampleRate: Double, targetDuration: TimeInterval = 0.8) {
        self.sampleRate = sampleRate
        self.targetDuration = targetDuration
    }

    public func reset() {
        pending.removeAll()
        nextSequence = 0
        bufferedSamples = 0
        started = false
    }

    /// Returns contiguous chunks once enough audio is buffered to start safely.
    /// Mid-session joiners often see high sequence numbers — we anchor to the
    /// earliest buffered sequence instead of waiting forever for sequence 0.
    public func append(header: AudioChunkHeader, pcm: Data) -> [Chunk] {
        guard pending[header.sequence] == nil else { return [] }
        if started && header.sequence < nextSequence { return [] }

        pending[header.sequence] = Chunk(header: header, pcm: pcm)
        bufferedSamples += Int(header.sampleCount)

        if !started {
            started = Double(bufferedSamples) / sampleRate >= targetDuration
            if started, let minSeq = pending.keys.min() {
                nextSequence = minSeq
            }
        }
        guard started else { return [] }

        var ready: [Chunk] = []
        while let chunk = pending.removeValue(forKey: nextSequence) {
            bufferedSamples -= Int(chunk.header.sampleCount)
            ready.append(chunk)
            nextSequence += 1
        }
        return ready
    }
}

File: android/core/sync/src/main/kotlin/com/chorus/core/sync/JitterBuffer.kt
package com.chorus.core.sync

import com.chorus.core.protocol.AudioChunkHeader

/**
 * Holds received PCM briefly so Wi-Fi jitter does not turn into audible gaps.
 * Reorders by sequence and releases a contiguous run once ~[targetDuration] is buffered.
 * Mid-session joiners anchor to the earliest buffered sequence (not zero).
 */
class AudioJitterBuffer(
    private val sampleRate: Double,
    private val targetDuration: Double = 0.8,
) {
    data class Chunk(val header: AudioChunkHeader, val pcm: ByteArray)

    private val pending = LinkedHashMap<Long, Chunk>()
    private var nextSequence: Long = 0
    private var bufferedSamples: Long = 0
    private var started = false

    fun reset() {
        pending.clear()
        nextSequence = 0
        bufferedSamples = 0
        started = false
    }

    fun append(header: AudioChunkHeader, pcm: ByteArray): List<Chunk> {
        val ready = ArrayList<Chunk>()
        if (pending.containsKey(header.sequence)) return ready
        if (started && header.sequence < nextSequence) return ready

        pending[header.sequence] = Chunk(header, pcm)
        bufferedSamples += header.sampleCount

        if (!started) {
            started = (bufferedSamples / sampleRate) >= targetDuration
            if (started) {
                nextSequence = pending.keys.minOrNull() ?: header.sequence
            }
        }
        if (!started) return ready

        while (true) {
            val chunk = pending.remove(nextSequence) ?: break
            bufferedSamples -= chunk.header.sampleCount
            ready.add(chunk)
            nextSequence++
        }
        return ready
    }
}
